### PR TITLE
JSUI-2562 OmniboxResultList receive event for querySuggest "selected" changes

### PR DESCRIPTION
**Pull request by orobert-coveo**
https://coveord.atlassian.net/browse/JSUI-2562

The first step to have a preview of the result when hovering over or moving with the arrow keys through the query suggestion.

This will allow the omniboxResultList to bind itself on an event when the "selected" querySuggest change 

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 1]
```diff
@@ -182,6 +202,9 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
         };
       });
     }
+    if (this.options.resultsPreview) {
+      //this.resultsPreview = new ResultsPreview()
```
Comment (samisayegh) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: Looks like this block can be removed?

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 2]
```diff
@@ -306,6 +336,23 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
...
+
+  private executeQueryHover(suggestion: string) {
+    let previousQueryOptions = this.queryController.getLastQuery();
+    previousQueryOptions.q = suggestion;
+    previousQueryOptions.numberOfResults = this.options.numberOfPreviewResult;
+    this.queryController
+      .getEndpoint()
+      .search(previousQueryOptions)
+      .then(results => {
+        //I now have the results for the preview.
```
Comment (samisayegh) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: Should there be some logic here? or will that be in a future PR?

Comment (orobert-coveo) [reply src/ui/OmniboxResultList/OmniboxResultList.ts]: in a future pr.
I'll remove it for this pr.

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 3]
```diff
@@ -152,10 +154,27 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
...
+     *
+     * this option will change from the default behavior of this component.
+     *
+     * When hovering on a query suggest, it will render the template receive to show a preview of the top results for this suggestion.
+     *
+     * **Default value:** `true`
+     */
+    resultsPreview: ComponentOptions.buildBooleanOption({ defaultValue: false }),
+    /**
+     * Number of results to render in the Omnibox.
```
[the remark this answers is not in the document]
Comment (orobert-coveo) [reply src/ui/OmniboxResultList/OmniboxResultList.ts]: Also, it is the number of results in the preview.
I will change it for something like :
`Number of results to render in the preview`

File: src/events/OmniboxEvents.ts [hunk 4]
```diff
@@ -80,4 +84,8 @@ export class OmniboxEvents {
    * This event is only triggered by standard ML-powered query suggestions, and not {@link AnalyticsSuggestions} or {@link FieldSuggestions}.
    */
   public static querySuggestSuccess = 'querySuggestSuccess';
+  /**
+   * Triggered by the [`Omnibox`]{@link Omnibox} component when a query suggestion is hovered by the mouse or selected with the keybord.
```
Comment (olamothe) [inline src/events/OmniboxEvents.ts]: `keyboard`

Also, this is pretty confusing, as the name is about "selection", but then it says it's triggered when a selection is hovered ?

There should probably be two different event: 

`querySuggestionHover`, when a suggestion is hovered through mouse or keyboard.
`querySuggestionSelection` when a suggestion is selected through mouse or keyboard.

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 5]
```diff
@@ -19,14 +19,16 @@ import OmniboxModuleDefintion = require('../Omnibox/Omnibox');
 import { InitializationEvents } from '../../EventsModules';
 import { logSearchBoxSubmitEvent } from '../Analytics/SharedAnalyticsCalls';
 import { Logger } from '../../misc/Logger';
-
 import 'styling/_OmniboxResultList';
+//import { ResultsPreview } from './ResultsPreview';
```
Comment (olamothe) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: Remove

File: src/events/OmniboxEvents.ts (hunk 4 above)
Comment (ThibodeauJF) [inline src/events/OmniboxEvents.ts]: keyboard*

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 6]
```diff
@@ -289,11 +301,15 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
     });
   }
 
-  private handleQueryOverride(args: IBuildingQueryEventArgs) {
+  protected handleBuildingQuery(args: IBuildingQueryEventArgs) {
     Assert.exists(args);
+    super.handleBuildingQuery(args);
```
Comment (olamothe) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: 🤔 ?

I think the way the code is right now, it's going to duplicate the work/ query expression...

File: src/ui/OmniboxResultList/OmniboxResultList.ts (hunk 5 above)
Comment (ThibodeauJF) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: Better not merge these comments right now

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 7]
```diff
@@ -306,6 +322,16 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
     }
   }
 
+  private handleQuerySuggestSelected(args: IQuerySuggestSelected) {
+    this.executeQueryHover(args.suggestion);
+  }
+
+  private executeQueryHover(suggestion: string) {
+    let previousQueryOptions = this.queryController.getLastQuery();
```
Comment (olamothe) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: `const`

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 8]
```diff
@@ -19,14 +19,16 @@ import OmniboxModuleDefintion = require('../Omnibox/Omnibox');
...
 import 'styling/_OmniboxResultList';
+//import { ResultsPreview } from './ResultsPreview';
 
 export interface IOmniboxResultListOptions extends IResultListOptions {
   omniboxZIndex?: number;
   onSelect?: (result: IQueryResult, resultElement: HTMLElement, omniboxObject: IPopulateOmniboxEventArgs, event?: Event) => void;
   headerTitle?: string;
   queryOverride?: IQueryExpression;
+  resultsPreview?: boolean;
+  numberOfPreviewResult?: number;
```
Comment (ThibodeauJF) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: Results should be plural

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 9]
```diff
@@ -306,6 +322,16 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
     }
   }
 
+  private handleQuerySuggestSelected(args: IQuerySuggestSelected) {
+    this.executeQueryHover(args.suggestion);
+  }
+
+  private executeQueryHover(suggestion: string) {
```
Comment (olamothe) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: I'm guessing this is incomplete ? As it does not actually seem like it execute anything meaningful, it just modifies an object without doing much with it.

Might be good to add some `TODO` in those case to help reviewers.

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 10]
```diff
@@ -306,6 +322,16 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
     }
   }
 
+  private handleQuerySuggestSelected(args: IQuerySuggestSelected) {
+    this.executeQueryHover(args.suggestion);
```
Comment (ThibodeauJF) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: Maybe you should not call this method unnecessarily (if the option `numberOfPreviewResult` is not defined by the user)

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 11]
```diff
@@ -152,7 +152,13 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
       (result: IQueryResult, resultElement: HTMLElement, omniboxObject: IPopulateOmniboxEventArgs) => void
     >(() => {
       return null;
-    })
+    }),
+    /**
+     * Number of results to render in the preview.
```
[the remark this answers is not in the document]
Comment (orobert-coveo) [reply src/ui/OmniboxResultList/OmniboxResultList.ts]: it doesn't affect the number of query suggestion we show so I'll change it for only the results
> The maximum number of query results to render in the preview.

File: src/ui/OmniboxResultList/OmniboxResultList.ts [hunk 12]
```diff
@@ -276,6 +283,9 @@ export class OmniboxResultList extends ResultList implements IComponentBindings
   }
 
   private handlePopulateOmnibox(args: IPopulateOmniboxEventArgs) {
+    if (this.options.numberOfPreviewResults > 0) {
```
Comment (samisayegh) [inline src/ui/OmniboxResultList/OmniboxResultList.ts]: I would store this condition in a getter, `shouldShowPreviewResults` and then use it here and on line 320.

**Pull request merged**